File: app/graphql/mutations/login_mutation.rb
class Mutations::LoginMutation < GraphQL::Schema::Mutation
  description 'Login to time clock'

  argument :username, String, required: true
  argument :password, String, required: true

  field :user, Types::UserType, null: false

  def resolve(username:, password:)
    unless context[:current_user_session]
      @user_session = UserSession.create({
        username: username,
        password: password
      })
      if @user_session.valid?
        { user: @user_session.user }
      else
        raise GraphQL::ExecutionError, 'Invalid login'
      end
    else
      raise GraphQL::ExecutionError, 'Not allowed'
    end
  end
end

File: app/javascript/components/mutations/loginMutation.js
import gql from 'graphql-tag';

export default gql`
  mutation login($username: String!, $password: String!) {
    loginMutation(username: $username, password: $password) {
      user {
        username
        level
        events {
          id
          kind
          comment
        }
      }
    }
  }
`;
